fix(NODE-6029): update types for collection listing indexes

prenaissance opened the pull request:
### Description

#### What is changing?

Update the return type of `Collection.prototype.indexes` and `Collection.prototype.indexInformation` to an array of indexes with full properties when the `full` option is true and to a Record of \<index name\> - \<keys and direction\> when the full option is false.

##### Is there new documentation needed for these changes?

No.

#### What is the motivation for this change?

The old types were incorrect in regards to the `full` option and were very wide. This will avoid making some type mistakes when using these methods.

<!--
Contributors!
First of all, thank you so much!!
If you haven't already, it would greatly help the team review this work in a timely manner if you create a JIRA ticket to track this PR.
You can do that here: https://jira.mongodb.org/projects/NODE
-->

### Release Highlight

<!-- RELEASE_HIGHLIGHT_START -->

### Index information helpers have accurate Typescript return types

`Collection.indexInformation()`, `Collection.indexes()` and `Db.indexInformation()` are helpers that return index information for a given collection or database.  These helpers take an option, `full`, that configures whether the return value contains full index descriptions or a compact summary:

```typescript
collection.indexes({ full: true });   // returns an array of index descriptions
collection.indexes({ full: false });  // returns an object, mapping index names to index keys
```

However, the Typescript return type of these helpers was always `Document`.  Thanks to @prenaissance, these helpers now have accurate type information!  The helpers return a new type,  `IndexDescriptionCompact | IndexDescriptionInfo[]`, which accurately reflects the return type of these helpers.  The helpers also support type narrowing by providing a boolean literal as an option to the API:

```typescript
collection.indexes();   // returns `IndexDescriptionCompact | IndexDescriptionInfo[]`
collection.indexes({ full: false });  // returns an `IndexDescriptionCompact`
collection.indexes({ full: true });  // returns an `IndexDescriptionInfo[]`

collection.indexInfo();   // returns `IndexDescriptionCompact | IndexDescriptionInfo[]`
collection.indexInfo({ full: false });  // returns an `IndexDescriptionCompact`
collection.indexInfo({ full: true });  // returns an `IndexDescriptionInfo[]`

db.indexInfo();   // returns `IndexDescriptionCompact | IndexDescriptionInfo[]`
db.indexInfo({ full: false });  // returns an `IndexDescriptionCompact`
db.indexInfo({ full: true });  // returns an `IndexDescriptionInfo[]`
```

<!-- RELEASE_HIGHLIGHT_END -->

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

baileympearson commented on src/operations/indexes.ts:
```diff
@@ -214,6 +214,18 @@ function resolveIndexDescription(
+ * @public
+ * The index information returned by the listIndexes command. https://www.mongodb.com/docs/manual/reference/command/listIndexes/#mongodb-dbcommand-dbcmd.listIndexes
+ */
+export type IndexDescriptionInfo = Omit<IndexDescription, 'key' | 'version'> & {
+  key: { [key: string]: IndexDirection };
+  v?: IndexDescription['version'];
+};
+
```
We generally try to keep our type definitions for server responses flexible so that if the server modifies a response in a future server version the driver is still compatible with it.  Could we `& Document` here?  I think that will provide type safety on known keys but will allow unknown keys without a Typescript error.

baileympearson commented on src/collection.ts:
```diff
@@ -693,8 +701,13 @@ export class Collection<TSchema extends Document = Document> {
    */
-  async indexInformation(options?: IndexInformationOptions): Promise<Document> {
-    return await this.indexes({ ...options, full: options?.full ?? false });
+  async indexInformation<TFull extends boolean = false>(
```
It's preferrable to avoid computed types as the return types from our APIs, since they don't provided users with information about the return type of a function until a user 1) reads the source of the computed type or 2) writes code using the API and checks the computed return type.  Could we accomplish the same here using overloads and removing all `TFull` generics?  Something like:

```typescript
  indexInformation(
    options: IndexInformationOptions & { full: true }
  ): Promise<IndexDescriptionInfo[]>;
  indexInformation(
    options: IndexInformationOptions & { full: false }
  ): Promise<IndexDescriptionCompact>;
  indexInformation(
    options: IndexInformationOptions & { full: boolean }
  ): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]>;
```

This has the additional benefits of 

- better documentation, because our tooling will generate a doc entry for each overload
- clearly defined overloads with existing IDE tooling
- each overload can be documented originally (if desired).

for example - here's what VSCode could hypothetically display if we used overloads:

<img width="793" alt="Screenshot 2024-04-09 at 12 23 01 PM" src="https://github.com/mongodb/node-mongodb-native/assets/23407842/ac95d9f5-07f1-4a29-984a-786eeb2bd2b6">
<img width="633" alt="Screenshot 2024-04-09 at 12 22 55 PM" src="https://github.com/mongodb/node-mongodb-native/assets/23407842/0e6c8042-c045-4f5f-8247-7440d3af9d46">

baileympearson commented on src/operations/indexes.ts:
```diff
@@ -72,7 +72,7 @@ export type IndexSpecification = OneOrMore<
 
 /** @public */
-export interface IndexInformationOptions extends ListIndexesOptions {
+export interface IndexInformationOptions<TFull extends boolean> extends ListIndexesOptions {
```
We could consider this a breaking change, since existing users' code that makes use of `IndexInformationOptions` would break because they wouldn't be providing a generic argument.  We could fix this by providing a default, but I think overloads are preferable (see my comment on `Collection.indexInformation()`).

baileympearson reviewed (changes requested):
Hey @prenaissance - thanks for the submission!  I left some comments about the implementation.

prenaissance commented:
> Hey @prenaissance - thanks for the submission! I left some comments about the implementation.

Thank you for the feedback. I fixed the comments and also added tests for `Db.prototype.indexInformation`

baileympearson commented on src/collection.ts:
```diff
@@ -121,6 +122,13 @@ export interface CollectionPrivate {
   writeConcern?: WriteConcern;
 }
 
+/** @public */
+export type IndexesInformation<TFull extends boolean = boolean> = TFull extends true
+  ? IndexDescriptionInfo[]
+  : TFull extends false
+  ? IndexDescriptionCompact
+  : IndexDescriptionInfo[] | IndexDescriptionCompact;
+
```
```suggestion
```

baileympearson replied:
(we'll also need to remove the export from the `index.ts`)

baileympearson commented on src/collection.ts:
```diff
@@ -693,8 +701,21 @@ export class Collection<TSchema extends Document = Document> {
+    options: IndexInformationOptions & { full: boolean }
+  ): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]>;
+  indexInformation(): Promise<IndexDescriptionCompact>;
+  async indexInformation(options?: IndexInformationOptions) {
```
```suggestion
  async indexInformation(options?: IndexInformationOptions): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]> {
```

We usually explicitly annotate the return type of all overloads, even when it can be inferred by TS.

baileympearson commented on src/collection.ts:
```diff
@@ -798,19 +819,23 @@ export class Collection<TSchema extends Document = Document> {
+    options: IndexInformationOptions & { full: boolean }
+  ): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]>;
+  indexes(): Promise<IndexDescriptionInfo[]>;
+  async indexes(options?: IndexInformationOptions) {
```
```suggestion
  async indexes(options?: IndexInformationOptions): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]> {
```

baileympearson commented on src/db.ts:
```diff
@@ -482,8 +484,23 @@ export class Db {
+    options: IndexInformationOptions & { full: boolean }
+  ): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]>;
+  indexInformation(name: string): Promise<IndexDescriptionCompact>;
+  async indexInformation(name: string, options?: IndexInformationOptions) {
```
```suggestion
  async indexInformation(name: string, options?: IndexInformationOptions): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]> {
```

prenaissance replied:
My bad!

baileympearson reviewed (changes requested):
one last round of small comments, otherwise LGTM

baileympearson commented on src/collection.ts:
```diff
@@ -693,8 +694,23 @@ export class Collection<TSchema extends Document = Document> {
+  indexInformation(
+    options: IndexInformationOptions & { full: boolean }
+  ): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]>;
+  indexInformation(): Promise<IndexDescriptionCompact>;
```
I was testing a bit more and I think we're missing one last overload.  Currently, the following has a type error:

```typescript
async function info(options: IndexInformationOptions) {
  const ret = await collection.indexInformation(options);
}

No overload matches this call.
  Overload 1 of 4, '(options: IndexInformationOptions & { full: true; }): Promise<IndexDescriptionInfo[]>', gave the following error.
    Argument of type 'IndexInformationOptions' is not assignable to parameter of type 'IndexInformationOptions & { full: true; }'.
      Type 'IndexInformationOptions' is not assignable to type '{ full: true; }'.
        Types of property 'full' are incompatible.
          Type 'boolean' is not assignable to type 'true'.
  Overload 2 of 4, '(options: IndexInformationOptions & { full?: false; }): Promise<IndexDescriptionCompact>', gave the following error.
    Argument of type 'IndexInformationOptions' is not assignable to parameter of type 'IndexInformationOptions & { full?: false; }'.
      Type 'IndexInformationOptions' is not assignable to type '{ full?: false; }'.
        Types of property 'full' are incompatible.
          Type 'boolean' is not assignable to type 'false'.
  Overload 3 of 4, '(options: IndexInformationOptions & { full: boolean; }): Promise<IndexDescriptionInfo[] | IndexDescriptionCompact>', gave the following error.
    Argument of type 'IndexInformationOptions' is not assignable to parameter of type 'IndexInformationOptions & { full: boolean; }'.
      Type 'IndexInformationOptions' is not assignable to type '{ full: boolean; }'.
        Property 'full' is optional in type 'IndexInformationOptions' but required in type '{ full: boolean; }'.ts(2769)
```

It seems like the problem is that the overload that matches a generic `IndexInformationOptions` contains a boolean literal (`{ full?: false }`), but because we don't have a const boolean, no overload is matched.

If we removed the `IndexInformationOptions & { full?: boolean }` overload and instead provided an overload:
```
indexInformation(
    options: IndexInformationOptions
  ): Promise<IndexDescriptionCompact | IndexDescriptionInfo[]>;
```

The issue is resolved.   This has the downside of losing the overload which provides a default return type when and empty options object is provided.

prenaissance replied:
I figured why this happens. This type `IndexInformationOptions & { full: boolean }` has the `full` property required and that makes `IndexInformationOptions` unassignable. Nice catch, I'll add a test case for this.

durran reviewed (approved):
I just updated the description from `Document.prototype` call outs to `Collection.prototype`
